Use IRGen's LinkInfo in TBDGen.

huonw opened the pull request:
Also, use it to generate symbols for conformance witness tables.

huonw committed:
[IRGen] Allow constructing UniversalLinkageInfo without an IRGenModule.

huonw committed:
[TBD] Thread enough through to use UniversalLinkageInfo.

huonw committed:
[TBDGen] Use LinkInfo for visibility and mangling, rather than reimplementing it.

huonw committed:
[TBDGen] Generate symbols for protocol conformances.

huonw commented:
@swift-ci Please smoke test.

slavapestov commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -761,10 +762,16 @@ static bool performCompile(std::unique_ptr<CompilerInstance> &Instance,
-        PrimarySourceFile ? validateTBD(PrimarySourceFile, *IRModule)
-                          : validateTBD(Instance->getMainModule(), *IRModule);
-    if (validationError)
+    auto hasMultipleIRGenThreads = Invocation.getSILOptions().NumThreads > 1;
```
This setting should not influence the set of public symbols we emit.

huonw replied:
I'm just passing through all the options that IRGen is asking for ( https://github.com/apple/swift/blob/d1c5de1cea6c388c78266cc1b691074d06b6eb33/lib/IRGen/GenDecl.cpp#L1372-L1386 ), to make sure we're matching as closely as possible. If you think this isn't required I'd be very happy to get rid of it.

slavapestov commented on lib/TBDGen/TBDGen.cpp:
```diff
@@ -83,36 +99,35 @@ class TBDGenVisitor : public ASTVisitor<TBDGenVisitor> {
+    addSymbol(irgen::LinkEntity::forTypeMetadataAccessFunction(declaredType));
+
+    // There are symbols associated with any protocols this type conforms to.
+    for (auto conformance : NTD->getAllConformances()) {
```
Isn't there a getLocalConformances() that only returns conformances defined in the same module?

slavapestov commented on lib/TBDGen/TBDGen.cpp:
```diff
@@ -35,23 +36,38 @@ static bool isPrivateDecl(ValueDecl *VD) {
 namespace {
 class TBDGenVisitor : public ASTVisitor<TBDGenVisitor> {
   StringSet &Symbols;
+  const irgen::UniversalLinkageInfo &UniversalLinkInfo;
```
using namespace swift::irgen; would help IMO

huonw committed:
[IRGen] Compute conformance fragility based on containing module.

Previously this would be using the current module's resilience strategy,
rather than that of the one that contains the conformance.

huonw committed:
[TBDGen] Say irgen::, irgen::, irgen::, irgen:: less.

huonw commented:
Updated for the review, except for the multiple threads point. And, also, includes a fix for the PublicExternal thing (first commit).

huonw commented:
@swift-ci Please smoke test.

slavapestov commented:
Looks good to me. If we need the multiple threads thing to initialize the IRGen it's fine, is just want to be extra careful that it doesn't change symbol visibility.
